hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemAttributes.java

```diff
@@ -42,70 +49,194 @@ public ITestAzureBlobFileSystemAttributes() throws Exception {
+      decodedAttributeValue2 = "Блюз"; //Блюз //你好
+      attributeValue1 = decodedAttributeValue1.getBytes(StandardCharsets.UTF_8);
+      attributeValue2 = decodedAttributeValue2.getBytes(StandardCharsets.UTF_8);
+    }
+    else {
+      decodedAttributeValue1 = "hi";
+      decodedAttributeValue2 = "hello"; // DFS Endpoint only Supports ASCII
+      attributeValue1 = fs.getAbfsStore().encodeAttribute(decodedAttributeValue1);
+      attributeValue2 = fs.getAbfsStore().encodeAttribute(decodedAttributeValue2);
+    }
+
+    // Attribute not present initially
+    assertNull(fs.getXAttr(path, attributeName1));
+    assertNull(fs.getXAttr(path, attributeName2));
+
+    // Set the Attributes
+    fs.setXAttr(path, attributeName1, attributeValue1);
+
+    // Check if the attribute is retrievable
+    byte[] rv = fs.getXAttr(path, attributeName1);
+    assertTrue(Arrays.equals(rv, attributeValue1));
+    assertEquals(new String(rv, StandardCharsets.UTF_8), decodedAttributeValue1);
+
+    // Set the second Attribute
+    fs.setXAttr(path, attributeName2, attributeValue2);
+
+    // Check all the attributes present and previous Attribute not overridden
+    rv = fs.getXAttr(path, attributeName1);
+    assertTrue(Arrays.equals(rv, attributeValue1));
+    assertEquals(new String(rv, StandardCharsets.UTF_8), decodedAttributeValue1);
+    rv = fs.getXAttr(path, attributeName2);
+    assertTrue(Arrays.equals(rv, attributeValue2));
+    assertEquals(new String(rv, StandardCharsets.UTF_8), decodedAttributeValue2);
+  }
+
+  /**
+   * Trying to set same attribute multiple times should result in no failure
+   * @throws Exception
+   */
+  @Test
+  public void testSetXAttrMultipleOperations() throws Exception {
+    AzureBlobFileSystem fs = getFileSystem();
+    final Path path = new Path("a/b");
+    fs.create(path);
+
+    String attributeName1 = "user.attribute1";
+    byte[] attributeValue1;
+    String decodedAttributeValue1 = "hi";
+
+    if(fs.getAbfsStore().getPrefixMode() == PrefixMode.BLOB) {
+      Assume.assumeTrue(!getIsNamespaceEnabled(fs));
+      attributeValue1 = "hi".getBytes(StandardCharsets.UTF_8);
+    }
+    else {
+      attributeValue1 = fs.getAbfsStore().encodeAttribute("hi");
+    }
+
+    // Attribute not present initially
+    assertNull(fs.getXAttr(path, attributeName1));
+
+    // Set the Attributes Multiple times
+    fs.setXAttr(path, attributeName1, attributeValue1);
+    fs.setXAttr(path, attributeName1, attributeValue1);
+
+    // Check if the attribute is retrievable
+    byte[] rv = fs.getXAttr(path, attributeName1);
+    assertTrue(Arrays.equals(rv, attributeValue1));
+    assertEquals(new String(rv, StandardCharsets.UTF_8), decodedAttributeValue1);
+  }
+
+  /**
+   * Test that setting metadata over marker blob do not override
+   * x-ms-meta-hdi_IsFolder
+   * TODO: Confirm Expected Behavior
+   * @throws Exception
+   */
+  @Test
+  public void testSetXAttrOverMarkerBlob() throws Exception {
```

Comment:
Also, we can add a test where in we force-set  x-ms-meta-hdi_IsFolder metadata on a normal blob. I observed it before that this doesn't get honored. That would be for future sanity.

Reply:
Strangely it is getting honored. I am able to set hdi_IsFolder metadata on a normal blob.

Reply:
Lets check with the wider forum (team) on how to go next on this.

Reply:
As discussed with the forum, we will keep this behavior as it is. Wasb is also allowing changing hdi_isFolder metadata.

Added a work item to take this up later.